### ISSUE TITLE
Support distributed trace propagation across SQS

### ISSUE THREAD

**Issue by randomanderson** [DataDog/dd-trace-java#1823]
SQS is a distributed queue by amazon.  It allows metadata with message attributes so implementation should be fairly straightforward.

**Comment by pedrorlmarques** [comment]
Do we have any estimation when this will be available ?

**Comment by randomanderson** [comment]
This quarter is the current target

**Comment by rud** [comment]
There has already been added, as I'm seeing SQS deliveries and retries under the original APM trace, as part of the same timeline. What is still missing from this?

**Comment by randomanderson** [comment]
> I believe this has been added, as I'm seeing SQS deliveries and retries under the original APM trace, as part of the same timeline. What is still missing from this?

We don't propagate headers over SQS so I'm not sure how its working for you

**Comment by randomanderson** [comment]
@rud Are you using JMS to interact with SQS?

**Comment by rud** [comment]
I am indeed using JMS to interact with SQS, maybe that is why it works for me. Sorry if I made things confusing.

**Comment by ADeane6** [comment]
@randomanderson Has there been any progress with propagating traces in SQS?, looking for a way to trace applications through SQS

**Comment by devinsba** [comment]
@ADeane6 are you using SDK v1 or v2? Any other information about your use case that could help us make sure we make it work for you would be appreciated!

**Comment by ADeane6** [comment]
@devinsba we are currently using SDK v1 in our kotlin services, we are using SNS fanout to SQS with other kotlin services consuming from the SQS queues. 
Ideally, we would like to propagate the trace ID through a message attribute and inject it into the trace context when the message is consumed.

**Comment by brentm5** [comment]
👋 Are there any updates on when this is expected to be added?

**Comment by mcculls** [comment]
propagation for native SQS and SNS is handled by #3104 via the `X-Amzn-Trace-Id` header - this includes injecting this header into outgoing AWS client requests and extracting it from AWS responses (or incoming server requests, for example when SNS triggers an endpoint that's being traced)

#3121 handles mapping the propagated `X-Amzn-Trace-Id` header to the appropriate JMS message attribute when consuming SQS messages as JMS

**Comment by hjizettle** [comment]
Is this done automatically? We are on version 0.95.1 of the agent and AWS SDK 1.12.165 and we are not seeing it flow across SQS.

**Comment by mcculls** [comment]
@hjizettle yes, it should be automatic when using the Java tracer - can you describe your flow (ie. is SNS involved at any point) and what languages are used on the producing and consuming sides

**Comment by hjizettle** [comment]
Yes we have SNS/SQS bridging happening to allow for reliable distribution to multiple consumers. Other than that there is nothing special going on really. We are a Java/Kotlin shop at least all the services involved here are either of these both consumers and producers.

**Comment by mcculls** [comment]
Is the SNS-SQS bridging being done in AWS? There have been reports of AWS not propagating the `X-Amzn-Trace-Id` header on their side when they fanout from SNS to SQS, and we're looking at ways to workaround this limitation.

**Comment by mcculls** [comment]
OK - in that case it would be best to handle this in a new issue about traces over SNS-SQS fanout, if you don't mind creating one. That way we can post updates for that scenario without disturbing everyone on this old issue.

The current workaround suggested by AWS is to use message headers which are propagated for SNS-SQS, it's unfortunate because in all other cases the `X-Amzn-Trace-Id` header is propagated by AWS - but for some reason it doesn't happen for SNS-SQS.

**Comment by armandobelardo** [comment]
@mcculls I'm seeing a similar issue with SQS requests being disjointed from the rest of the trace (ie the request that produces the message and ultimately the consumer), we're using the SDK directly (`v2.17.122`), with the Java tracer at `v0.97.0` without SNS involved.
Is there a suggested way to debug to see if I could get more information on where the stoppage is?
If it's helpful, we're seeing AWS MSK consumers and produces within the DD APM without issue.

**Comment by phamlehieu** [comment]
@mcculls I'm using [Spring Cloud AWS](https://spring.io/projects/spring-cloud-aws) and it doesn't work for my case. Is there any additional setup to enable the sqs propagation?
![example_trace](https://user-images.githubusercontent.com/6805429/173210775-405127a2-a77c-456d-9809-8b4d53a712b7.png)

FYI: [Spring Cloud AWS](https://spring.io/projects/spring-cloud-aws) use AsyncSQSClient of the AWS SDK v1

**Comment by scottberres-tr** [comment]
. 

> @mcculls I'm using [Spring Cloud AWS](https://spring.io/projects/spring-cloud-aws) and it doesn't work for my case. Is there any additional setup to enable the sqs propagation? ![example_trace](https://user-images.githubusercontent.com/6805429/173210775-405127a2-a77c-456d-9809-8b4d53a712b7.png)
> 
> FYI: [Spring Cloud AWS](https://spring.io/projects/spring-cloud-aws) use AsyncSQSClient of the AWS SDK v1

We are seeing the same thing on our end. Passing all of the required headers in the message attributes does not propagate the data dog trace id. I have validated on the receiving side that those headers are there, but datadog tracing is not propagated to the receiving side and any other calls beyond that. Is datadog trace supported in this case, and if so, what needs to be done to propagate the datadog trace?  Thanks

**Comment by phamlehieu** [comment]
Maybe it related to https://github.com/DataDog/dd-trace-java/issues/3861 @scottberres-tr

**Comment by mcculls** [comment]
If you're seeing trace attributes in the received messages but a trace still isn't being activated then unfortunately you will need to manually activate it. This can be done using the OpenTracing API - you'll need to extract the trace details and activate a span using the extracted context as show in https://docs.datadoghq.com/tracing/trace_collection/open_standards/java/#inject-and-extract-context-for-distributed-tracing.

For further updates relating to automatic SQS propagation please follow issue https://github.com/DataDog/dd-trace-java/issues/3411 - I'll let you know when we have a new snapshot to test with SQS.

**Comment by mcculls** [comment]
The Datadog Java tracer will create `AWSTraceHeader` system attributes for any message sent via the SQS client API. It then relies on AWS to propagate those attributes to the receiving end.

If you're seeing the `AWSTraceHeader` system attribute in your received message, but you're not seeing a trace activated using that context (in other words the trace is still disconnected) then you'll need to activate it manually. This activation happens automatically when using clients such as SQS-JMS, but other clients need additional instrumentation and for them manual activation is the recommended workaround.

The easiest way to activate the received `AWSTraceHeader` is by using the OpenTracing API to extract the trace id and parent span id from the header and build a new span using that context which you then activate.

Otherwise you'll need to wait for the next iteration of the SQS instrumentation, which will extend the instrumentation so this workaround is not required - there will be an update on https://github.com/DataDog/dd-trace-java/issues/3411 when it's available for testing
